src/main/java/hudson/plugins/s3/S3BucketPublisher.java

```diff
@@ -283,7 +293,12 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath ws, @Nonnull Launc
             }
         } catch (AmazonClientException|IOException e) {
             e.printStackTrace(listener.error("Failed to upload files"));
-            run.setResult(constrainResult(Result.UNSTABLE, listener));
+            if (!isDontSetBuildResultOnFailure()) {
+                run.setResult(constrainResult(Result.UNSTABLE, listener));
+            } else {
+                throw new IOException("Failed to upload files");
```

Comment:
Please, attach the source exception to be presented to the users so they can investigate.